src/devices/src/virtio/block/device.rs

```diff
@@ -332,886 +332,627 @@ impl VirtioDevice for Block {
+        // Write.
+        {
+            mem.write_obj_at_addr::<u32>(VIRTIO_BLK_T_OUT, request_type_addr)
+                .unwrap();
+            // Make data read only, 8 bytes in len, and set the actual value to be written.
+            vq.dtable[1].flags.set(VIRTQ_DESC_F_NEXT);
+            vq.dtable[1].len.set(8);
+            mem.write_obj_at_addr::<u64>(123_456_789, data_addr)
+                .unwrap();
+
+            check_metric_after_block!(
+                &METRICS.block.write_count,
+                1,
+                invoke_handler_for_queue_event(&mut block)
+            );
+
+            assert_eq!(vq.used.idx.get(), 1);
+            assert_eq!(vq.used.ring[0].get().id, 0);
+            assert_eq!(vq.used.ring[0].get().len, 0);
+            assert_eq!(
+                mem.read_obj_from_addr::<u32>(status_addr).unwrap(),
+                VIRTIO_BLK_S_OK
+            );
+        }
+
+        // Read.
+        {
+            vq.used.idx.set(0);
+            block.set_queue(0, vq.create_queue());
+
+            mem.write_obj_at_addr::<u32>(VIRTIO_BLK_T_IN, request_type_addr)
+                .unwrap();
+            vq.dtable[1]
+                .flags
+                .set(VIRTQ_DESC_F_NEXT | VIRTQ_DESC_F_WRITE);
+
+            check_metric_after_block!(
+                &METRICS.block.read_count,
+                1,
+                invoke_handler_for_queue_event(&mut block)
+            );
+
+            assert_eq!(vq.used.idx.get(), 1);
+            assert_eq!(vq.used.ring[0].get().id, 0);
+            assert_eq!(vq.used.ring[0].get().len, vq.dtable[1].len.get());
+            assert_eq!(
+                mem.read_obj_from_addr::<u32>(status_addr).unwrap(),
+                VIRTIO_BLK_S_OK
+            );
+            assert_eq!(
+                mem.read_obj_from_addr::<u64>(data_addr).unwrap(),
+                123_456_789
+            );
+        }
+    }
+
+    #[test]
+    fn test_flush() {
+        let mut block = default_block();
+        let mem = block.mem.clone();
+        let vq = VirtQueue::new(GuestAddress(0), &mem, 16);
+        block.set_queue(0, vq.create_queue());
+        block.set_device_activated(true);
+        initialize_virtqueue(&vq);
+
+        let request_type_addr = GuestAddress(vq.dtable[0].addr.get());
+        let status_addr = GuestAddress(vq.dtable[2].addr.get());
+
+        // Flush completes successfully when data descriptor is provided.
+        {
+            mem.write_obj_at_addr::<u32>(VIRTIO_BLK_T_FLUSH, request_type_addr)
+                .unwrap();
+
+            invoke_handler_for_queue_event(&mut block);
+            assert_eq!(vq.used.idx.get(), 1);
+            assert_eq!(vq.used.ring[0].get().id, 0);
+            assert_eq!(vq.used.ring[0].get().len, 0);
+            assert_eq!(
+                mem.read_obj_from_addr::<u32>(status_addr).unwrap(),
+                VIRTIO_BLK_S_OK
+            );
+        }
+
+        // Flush completes successfully even without a data descriptor.
```

Comment:
This is the default. The flush shouldn't really have a data descriptor. For now we can just remove "even without" from this comment, but in the future we might want to actually ignore flush requests where the data descriptor is present.

Reply:
Right, then I guess the correct way to put it is "it works even with a data descriptor".